fdbserver/storageserver.actor.cpp

```diff
@@ -3360,6 +3361,36 @@ bool storageServerTerminated(StorageServer& self, IKeyValueStore* persistentData
 		return false;
 }
 
+ACTOR Future<bool> memoryStoreRecover(IKeyValueStore* store, Database cx, UID id)
+{
+    if(store->getType() != KeyValueStoreType::MEMORY) {
+        return false;
+    }
+
+    state Transaction tr( cx );
+	state int noCanRemoveCount = 0;
+	loop {
+		try {
+			tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+
+			state bool canRemove = wait( canRemoveStorageServer( &tr, id ) );
+			if (!canRemove) {
+				TEST(true); // it's possible that the caller had a transaction in flight that assigned keys to the server. Wait for it to reverse its mistake.
+				Void _ = wait( delayJittered(SERVER_KNOBS->REMOVE_RETRY_DELAY, TaskUpdateStorage) );
+				tr.reset();
+				TraceEvent("RemoveStorageServerRetrying").detail("Count", noCanRemoveCount++).detail("ServerID", id).detail("CanRemove", canRemove);
+			} else {
+				Void _ = wait(tr.commit());
+				return true;
+			}
+		} catch (Error& e) {
+			state Error err = e;
+			Void _ = wait( tr.onError(e) );
```

Comment:
I think `tr.onError` could throw in some cases? I think you would probably want to make sure this actor never throws. I probably would do a simple hack like this:

 ```suggestion
			try { Void _ = wait( tr.onError(e) ); } catch (Error& e) { /* insert a TraceEvent here? */ return Never(); }
```